src/sensors/elec.rs

```diff
@@ -14,21 +26,142 @@ pub fn filter_pde(pde: f32) -> bool {
+/// might prove too inflexible. Should be generalised.
+pub fn combine_sensor_hits(event_id: u32, hits: Vec<MCSensorHit>) -> Vec<MCQT> {
+    hits.into_iter()
+        .group_by(|h| h.sensor_id)
+        .into_iter()
+        .map(|(s, grp)| qt_min_time(event_id, s, grp.collect()))
+        .collect()
+}
+
+// This seems like something that would be in a crate somewhere.
+fn dot((x1,y1,z1): (Length, Length, Length), (x2,y2,z2): (Length, Length, Length)) -> Area {
+    x1*x2 + y1*y2 + z1*z2
+}
+
+// Dot product separation for now, need to implement other algorithm(s)
+fn dot_product_clusters(hits: &[MCQT], xyzs: &SensorMap, threshold: f32, min_sensors: usize) -> Option<(Vec<MCQT>, Vec<MCQT>)> {
+    let max_sens = hits.iter().max_by_key(|e| e.q).map(|e| e.sensor_id)?;
+    let &(xm, ym, zm) = xyzs.get(&max_sens)?;
+    let (c1, c2): (Vec<MCQT>, Vec<MCQT>) = hits.iter()
+            .filter(|&MCQT { q, .. }| *q as f32 > threshold)
+            .partition(|&MCQT { sensor_id, .. }| dot((xm, ym, zm), *xyzs.get(sensor_id).unwrap()) > Area::ZERO);
+    if c1.len() >= min_sensors && c2.len() >= min_sensors { Some((c1, c2)) } else { None }
+}
+
+/// Reconstruct LORs based on the barycentre of the 2 selected clusters (dot prod) without DOI.
+fn lor_from_sensor_positions(hits: &[MCQT], xyzs: &SensorMap, threshold: f32, min_sensors: usize) -> Option<Hdf5Lor> {
+    let (c1, c2) = dot_product_clusters(hits, xyzs, threshold, min_sensors)?;
+    // TODO time just taken as minimum of minima. Averaging used elsewhere.
+    let b1 = sipm_charge_barycentre(&c1, xyzs)?;
+    let b2 = sipm_charge_barycentre(&c2, xyzs)?;
+    let q1 = c1.iter().map(|&MCQT { q, .. }| q as f32).sum();
+    let q2 = c2.iter().map(|&MCQT { q, .. }| q as f32).sum();
+    // Just dummy time for now until we work out the best way to sort.
+    Some(Hdf5Lor {
+        dt: ns_(b2.3 - b1.3),
+        x1: mm_(b1.0), y1: mm_(b1.1), z1: mm_(b1.2),
+        x2: mm_(b2.0), y2: mm_(b2.1), z2: mm_(b2.2),
+        q1, q2, E1: f32::NAN, E2: f32::NAN,
+    })
+}
+
+fn sipm_charge_barycentre(hits: &[MCQT], xyzs: &SensorMap) -> Option<(Length, Length, Length, Time)> {
+    let mut qs = Ratio::ZERO;
+    let mut xx = Length::ZERO;
+    let mut yy = Length::ZERO;
+    let mut zz = Length::ZERO;
+    let mut tt = Time::ZERO;
+    for &MCQT{ sensor_id, q, t, .. } in hits {
+        let &(x, y, z) = xyzs.get(&sensor_id)?;
+        let q = ratio(q as f32);
+        qs += q;
+        xx += x * q;
+        yy += y * q;
+        zz += z * q;
+        if t < tt { tt = t};
+    }
+    Some((xx / qs, yy / qs, zz / qs, tt))
+}
+
+
+pub fn lor_reconstruction(first_file: &PathBuf, pde: f32, time_mu: f32, time_sigma: f32, threshold: f32, min_sensors: usize) -> Box<dyn Fn(&PathBuf) -> hdf5::Result<(Vec<Hdf5Lor>, usize)>> {
+    let time_smear = gaussian_sampler(time_mu, time_sigma);
+    // Need to do better than unwrap!
+    let xyzs = read_sensor_map(&first_file).unwrap();
```

Comment:
Right, the `usize` is part of the `Ok` variant, and it's been absorbed by the new `LorBatch` type.

So I think you should simply replace the `.unwrap()` with `?`.

Reply:
So I still can't use the `?` because I define the variable in the wrapping function rather than in the returned closure. I did it that way to not redefine each time the function is called, is that an issue in `RUST`? Or am I thinking pythonly?

Reply:
Ah, sorry, I misread the return type: I thought it was 
```rust
hdf5:::Result<_, _>
```
but it's actually
```rust
 Box<dyn Fn(&PathBuf) -> hdf5::Result<(_, _)>>
```
Generally speaking, `x?` is just shorthand for

```rust
match x {
    Ok(v)  => v,
    Err(e) => return Err(e),
};
```    
or
```rust
match x {
    Some(v)  => v,
    None     => return None,
};
```
That is to say, it unwraps the value (if `Some` or `Ok`) or returns the failure (if `None` or `Err`).

So you can use it after *any* expression whose type is `Option<_>` or `Result<_, _>` and matches the return type of the function immediately surrounding the `?`.

Back to your specific problem: just pass in `xyzs` instead of `first_file`: `xysz` is already defined in the scope from which you call `lor_reconstruction`.

> I did it that way to not redefine each time the function is called, is that an issue in RUST? Or am I thinking pythonly?

Not sure what you mean. You mean defining `xyzs` in the closure? That would result in `xyzs` being initialized every time the closure is called. Your current solution initializes `xysz` every time `lor_reconstruction` is called (hopefully only once). My proposed solution initializes `xyzs` exactly once.

Reply:
> Back to your specific problem: just pass in xyzs instead of first_file: xysz is already defined in the scope from which you call lor_reconstruction.

I'm running into lifetime problems with the borrow. I can get the function itself to be happy with either `'static` lieftime for the `SensorMap` parameter or a specific lifetime with a bound for the closure but in both cases the call in `makelor.rs` wants `'static`.

Reply:
See the two commits I just pushed to your branch.

Reply:
Yep, thanks. Stupid error, I was forgetting the change to the `FilenameToLorsFunction` to add the lifetime.

Reply:
I forgot too, but the compiler reminded me.

(Even though Rust's compiler errors are spectacularly better than other languages', all too often it's still difficult to find the signal among the noise. This case wasn't so bad.)